Using TransactionAware via FlowletContext

ParthGandhi opened the pull request:
Added ability to use `TransactionAware`s in Flowlets via the `FlowletContext`. 

Added a `ForwardingFlowletContext` that simply delegates to the underlying FlowletContext, but also adds TxAwares to the current transaction. 
Removed `createTransactionExecutor()` in `DataFabricFacade` in favor of manually dealing with the transactional lifecycle.This allows us to add TxAwares dynamically.

Commit:
Ability to add TransactionAware via FlowletContext. Relevant tests.

gokulavasan commented on tigon-flow/pom.xml:
```diff
@@ -53,10 +53,22 @@ the License.
     </dependency>
+    <dependency>
+      <groupId>com.continuuity.tephra</groupId>
+      <artifactId>tephra-hbase-compat-0.94</artifactId>
```
Does it need dependency on tephra-hbase-compat-0.96 as well?

ParthGandhi replied:
Fixed this to be only under "test" scope. This is needed to use `TransactionAwareHTable` in the unit-tests.

chtyim replied:
For test, actually it's better to use the 0.96 compat module.

Commit:
Fix checkstyle and maven build

achalpandey commented on tigon-api/pom.xml:
```diff
@@ -48,6 +48,10 @@
+    </dependency>
   </dependencies>
 
 </project>
```
Add newline ?

Commit:
Use hbase-0.96 for test

ParthGandhi commented:
https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT2-3

achalpandey commented on tigon-flow/src/test/java/co/cask/tigon/flow/TransactionAwareFlowTest.java:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * Test Flow that writes to an instance of TransactionAware.
+   */
+  public static final class TestFlow implements Flow {
```
Can we reuse the same class from https://github.com/caskco/tigon/blob/feature/bigflowapi/tigon-flow/src/test/java/co/cask/tigon/flow/test/BasicFlowTest.java#L63

Same for GeneratorFlowlet and SinkFlowlet

ParthGandhi replied:
No. I'm using that class in my integration test as part of https://github.com/caskco/tigon/pull/15. They do entirely different things, and thus it doesn't make sense to reuse it in both test classes.

achalpandey replied:
Understood.
[Adding comment for completeness]

achalpandey commented on tigon-flow/src/test/java/co/cask/tigon/flow/TransactionAwareFlowTest.java:
```diff
@@ -0,0 +1,189 @@
+    }
+  }
+
+  private static final class GeneratorFlowlet extends AbstractFlowlet {
```
Reuse https://github.com/caskco/tigon/blob/feature/bigflowapi/tigon-flow/src/test/java/co/cask/tigon/flow/test/BasicFlowTest.java#L79

achalpandey commented on tigon-flow/src/test/java/co/cask/tigon/flow/TransactionAwareFlowTest.java:
```diff
@@ -0,0 +1,189 @@
+    }
+  }
+
+  private static final class SinkFlowlet extends AbstractFlowlet {
```
Reuse https://github.com/caskco/tigon/blob/feature/bigflowapi/tigon-flow/src/test/java/co/cask/tigon/flow/test/BasicFlowTest.java#L115

Commit:
Merge dev

sreevatsanraman commented on tigon-flow/src/test/java/co/cask/tigon/flow/TransactionAwareFlowTest.java:
```diff
@@ -0,0 +1,189 @@
+  public void testAddTransactionAware() throws Exception {
+    FlowManager flowManager = deployFlow(TestFlow.class, ImmutableMap.<String, String>of());
+
+    TimeUnit.SECONDS.sleep(5);
```
This might fail in bamboo if the machine is slow. Need to check for status at smaller intervals (1 sec) and timeout if the status is not as expected after say 60 seconds

ParthGandhi replied:
As far as I know, we don't actually expose "status" to a user. So this doesn't seem feasible at this point. 

@gokulavasan @achalpandey Is there a smarter way of doing this that I'm missing?

sreevatsanraman commented on tigon-flow/src/test/java/co/cask/tigon/flow/TransactionAwareFlowTest.java:
```diff
@@ -0,0 +1,189 @@
+
+    flowManager.stop();
+
+    TimeUnit.SECONDS.sleep(5);
```
Same here.

chtyim commented on pom.xml:
```diff
@@ -100,6 +100,11 @@
       </dependency>
       <dependency>
         <groupId>com.continuuity.tephra</groupId>
```
Use the co.cask one. Although you'll need to temporarily add our internal repo to the repositories section in the pom.

chtyim commented on pom.xml:
```diff
@@ -100,6 +100,11 @@
+        <version>${tephra.version}</version>
+      </dependency>
+      <dependency>
+        <groupId>com.continuuity.tephra</groupId>
```
same here.

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/BasicFlowletContext.java:
```diff
@@ -107,6 +123,10 @@ public String getFlowletId() {
   }
 
+  public List<TransactionAware> getTransactionAwares() {
+    return transactionAwares;
```
Wrap it with a unmodifiable view of it. Or, better yet, not exposing this method, but rather have it like:

`void addTransactionAwares(TransactionContext ctx);`

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/FlowletProcessDriver.java:
```diff
@@ -370,37 +372,44 @@ public void ack() throws TransactionFailureException {
+        LOG.info("Flowlet initialized: " + flowletContext);
+
+        txContext.finish();
+      } catch (Exception e) {
```
This is not exactly the same as what TransactionExecutor does. It does it like:

``` java
try {
  txContext.start();
  try {
    // call user code
  } catch (Throwable t) {
    txContext.abort(new TransactionFailureException("", t));  // This will throw exception again.
  }
  txContext.finish();   // This will throw if fails to commit
} catch (TransactionFailureException e) {
  ...
}
```

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/ForwardingFlowletContext.java:
```diff
@@ -0,0 +1,82 @@
+ * A forwarding {@link FlowletContext} that delegates to the underlying instance.
+ * It adds any {@link TransactionAware}s to the current {@link TransactionContext}.
+ */
+public class ForwardingFlowletContext implements FlowletContext {
```
Usually a Forwarding class is an abstract class.

chtyim replied:
Looking at what this class does, it is actually not a pure Forwarding class. In fact, can it be just part of BasicFlowletContext?

Commit:
Addressing comments

Commit:
Invert adding of txAwares

Commit:
Addressing comments

Commit:
Checkstyle

ParthGandhi commented:
@chtyim Addressed comments. Please review. 

Got rid of `ForwardingFlowletContext`. In place, anytime a `TransactionAware` is added to the `flowletContext`, it is also added to the current `transactionContext`. Also, whenever a new `TransactionContext` it set, any `TransactionAwares` that have been previously added are added to the context too.

Commit:
Merge branch 'develop' of github.com:caskco/tigon into feature/context-txaware

Conflicts:
	pom.xml
	tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/AbstractDataFabricFacade.java
	tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/FlowletProcessDriver.java

Commit:
Switch to co.cask.tephra

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/AbstractDataFabricFacade.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2014 Cask Data, Inc.
+ * Copyright 2014 Cask Data, Inc.
```
This is wrong change, right?

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/FlowletProcessDriver.java:
```diff
@@ -257,9 +257,10 @@ public void run() {
+    final TransactionContext txContext = dataFabricFacade.createTransactionManager();
     try {
       txContext.start();
+      flowletContext.setTransactionContext(txContext);
```
Better set before calling start.

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/FlowletProcessDriver.java:
```diff
@@ -370,37 +371,45 @@ public void ack() throws TransactionFailureException {
-        }
-      });
+      txContext.start();
+      flowletContext.setTransactionContext(txContext);
```
set before start.

chtyim commented:
Almost good. Just couple comments left.

Commit:
Addressed comments

Commit:
Missed setting txcontext properly

ParthGandhi commented:
@chtyim Please review again.

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/FlowletProcessDriver.java:
```diff
@@ -370,37 +371,45 @@ public void ack() throws TransactionFailureException {
-          LOG.info("Flowlet initialized: " + flowletContext);
-        }
-      });
+      flowletContext.setTransactionContext(txContext);
```
hum.... now reading the code again makes me wonder, shouldn't it be the txContext created by the BasicFlowletContext?

``` java
class BasicFlowletContext {
...
  TransactionContext createTxContext() {
    // create, add txawares, return.
  }
}
```

Commit:
Merge branch 'develop' of github.com:caskco/tigon into feature/context-txaware

Conflicts:
	pom.xml

Commit:
Address comments

ParthGandhi commented:
@chtyim Agreed. I think that makes more sense too. 

Please review again.

Commit:
Merge dev

chtyim commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/flow/BasicFlowletContext.java:
```diff
@@ -87,6 +99,20 @@ public FlowletSpecification getSpecification() {
+  @Override
+  public void addTransactionAware(TransactionAware transactionAware) {
+    transactionAwares.add(transactionAware);
+    transactionContext.addTransactionAware(transactionAware);
```
transactionContext could be null.

chtyim commented:
Just one comment. Fix it then it's good to go.

Commit:
Merge branch 'develop' of github.com:caskco/tigon into feature/context-txaware

Commit:
Null check for adding txAwares

Commit:
Merge branch 'develop' of github.com:caskco/tigon into feature/context-txaware

chtyim commented:
LGTM. Merge it.

Commit:
Fix pom

Commit:
Fix pom

Commit:
Remove useless file

Commit:
Ignore failing test

Commit:
Remove unused file

Commit:
Put back file

ParthGandhi commented:
https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT28-1
